server/src/main/java/org/elasticsearch/search/aggregations/metrics/CardinalityAggregatorFactory.java

```diff
@@ -48,7 +49,28 @@
     }
 
     public static void registerAggregators(ValuesSourceRegistry.Builder builder) {
-        builder.register(CardinalityAggregationBuilder.REGISTRY_KEY, CoreValuesSourceType.ALL_CORE, CardinalityAggregator::new, true);
+        builder.register(CardinalityAggregationBuilder.REGISTRY_KEY,
+            CoreValuesSourceType.ALL_CORE,
+            (name, valuesSourceConfig, precision, context, parent, metadata) -> {
+                // super hacky but it shows the point of the approach
+                if (valuesSourceConfig.hasValues()) {
+                    ValuesSource valuesSource = valuesSourceConfig.getValuesSource();
+                    if (valuesSource instanceof ValuesSource.Bytes.WithOrdinals) {
+                        ValuesSource.Bytes.WithOrdinals source = (ValuesSource.Bytes.WithOrdinals) valuesSource;
+                        // is this call expensive? I hope not
+                        final long maxOrd = source.globalMaxOrd(context.searcher());
+                        final long countsMemoryUsage = HyperLogLogPlusPlus.memoryUsage(precision);
+                        final long ordinalsMemoryUsage = maxOrd * 4;
+                        // better heuristic? Currently we pay the price of the bitSet but it should be small
+                        if (ordinalsMemoryUsage < countsMemoryUsage) {
```

Comment:
It can be quite large if the incoming cardinality is high and the cardinality of the field we're measuring is low. We have that problem in the implementation we have today though. I mean, like, say you have `date_hist(hours) -> terms(high_cardinality) -> cardinality(average value around 2 or something)`. In that case we'll allocate the whole bit map for every incoming ord, only to set one or two bits.